File: WoofAndroid/konsist/src/test/kotlin/com/perrystreet/woof/konsist/designsystem/UIComponentsDoNotUseHardcodedValues.kt
package com.perrystreet.woof.konsist.designsystem

import com.lemonappdev.konsist.api.ext.list.withAnnotationNamed
import com.lemonappdev.konsist.api.ext.list.withoutAnnotationNamed
import com.perrystreet.woof.konsist.Assertions.assertFalse
import com.perrystreet.woof.konsist.KonsistUtils
import com.perrystreet.woof.konsist.LintRuleMessage
import io.kotest.core.spec.style.BehaviorSpec

class UIComponentsDoNotUseHardcodedValues : BehaviorSpec() {
    init {
        Given("A composable outside the primitives") {
            val composables = KonsistUtils.productionCode.functions()
                .withAnnotationNamed("Composable")
                .withoutAnnotationNamed("PreviewDevices")
                .filter { !it.path.contains("/_primitives/") }

            Then("It does not hardcode dp values") {
                composables.assertFalse(message = DpMessage) { HardcodedDpRegex.containsMatchIn(it.text) }
            }

            Then("It does not hardcode colors") {
                composables.assertFalse(message = ColorMessage) { HardcodedColorRegex.containsMatchIn(it.text) }
            }
        }
    }

    private companion object {
        private val HardcodedDpRegex = Regex("""[1-9]\d*\.dp\b""")
        private val HardcodedColorRegex = Regex("""Color\(0x""")

        private val DpMessage = LintRuleMessage(
            rule = "Composables never hardcode dp values.",
            why = "Spacing, sizing, and radii are tokens tuned in one place. A literal dp bypasses that and drifts between components.",
            howToFix = "Use Theme.padding.*, Theme.sizing.*, SpacingRoles.*, or SizingRoles.* instead of the literal.",
            badExample = "Modifier.padding(16.dp)",
            goodExample = "Modifier.padding(Theme.padding.elementRegular)",
        )

        private val ColorMessage = LintRuleMessage(
            rule = "Composables never hardcode colors.",
            why = "Colors come from the theme so light and dark variants stay in sync. A literal color ignores the active theme.",
            howToFix = "Add a semantic color to Colors and read it through Theme.colors or a *ColorRole.",
            badExample = "Text(color = Color(0xFFFF7A29))",
            goodExample = "AtomText(colorRole = TextColorRole.Primary)",
        )
    }
}

File: WoofIOS/SwiftPackages/HarmonizeRules/Tests/HarmonizeRulesTests/DesignSystem/UIComponentsDoNotUseHardcodedValues.swift
import Foundation
import Harmonize
import HarmonizeSemantics
import Quick

final class UIComponentsDoNotUseHardcodedValues: QuickSpec {
    override class func spec() {
        Given("A view outside the primitives") {
            let views = WoofHarmonize.views.withoutFolder("_Primitives")

            Then("It does not hardcode point values") {
                views.assertFalse(message: pointMessage) { $0.description.containsMatch(of: hardcodedPointPattern) }
            }

            Then("It does not hardcode colors") {
                views.assertFalse(message: colorMessage) { $0.description.containsMatch(of: hardcodedColorPattern) }
            }

            Then("It does not build static gradients inline") {
                views
                    .filter { !$0.description.contains("TimelineView") }
                    .assertFalse(message: gradientMessage) { $0.description.containsMatch(of: inlineGradientPattern) }
            }
        }
    }

    private static let hardcodedPointPattern = try! NSRegularExpression(
        pattern: #"\b(padding|spacing|cornerRadius|radius|width|height|minHeight|maxHeight|minWidth|maxWidth|offset|x|y)\s*:\s*-?[1-9]\d*(\.\d+)?\b|\.padding\(\s*[1-9]\d*|\.frame\(\s*[1-9]\d*"#
    )
    private static let hardcodedColorPattern = try! NSRegularExpression(pattern: #"Color\((hex|red|white|hue):"#)
    private static let inlineGradientPattern = try! NSRegularExpression(pattern: #"\b(Linear|Radial|Angular|Elliptical)Gradient\s*\("#)

    private static let pointMessage = LintRuleMessage(
        rule: "Views never hardcode point values.",
        why: "Spacing, sizing, and radii are tokens tuned in one place. A literal point value bypasses that and drifts between components.",
        howToFix: "Use PaddingRoles.*, SpacingRoles.*, SizingRoles.*, or theme.radius.* instead of the literal.",
        badExample: ".padding(16)",
        goodExample: ".padding(PaddingRoles.Element.regular.rawValue)"
    )

    private static let colorMessage = LintRuleMessage(
        rule: "Views never hardcode colors.",
        why: "Colors come from the theme so light and dark variants stay in sync. A literal color ignores the active theme.",
        howToFix: "Add a semantic color to Colors and read it through theme.colors or a *ColorRole.",
        badExample: "Text(text).foregroundStyle(Color(hex: 0xFF7A29))",
        goodExample: "AtomText(text: text, textFontRole: .bodyP1, colorRole: .primary)"
    )

    private static let gradientMessage = LintRuleMessage(
        rule: "Views never build static gradients inline.",
        why: "A gradient is a design decision: which colors, and where the ramp starts. Built inline it drifts between components, and its stop positions become literals no token can tune. Animated gradients driven by TimelineView are part of their component's motion and stay private to it.",
        howToFix: "Add the gradient to GradientRoles and read it through theme.gradients.",
        badExample: "LinearGradient(colors: [Color.clear, theme.colors.scrimDim], startPoint: .top, endPoint: .bottom)",
        goodExample: "theme.gradients.scrimVertical"
    )
}
